File: kotlin/src/test/java/cn/staynoob/trap/java/InterfaceSpec.java
package cn.staynoob.trap.java;

import cn.staynoob.trap.java.interfacespec.IWithMethodImplementation1;
import cn.staynoob.trap.java.interfacespec.IWithMethodImplementation2;
import org.junit.Test;

import static org.assertj.core.api.Assertions.*;

public class InterfaceSpec {

    class Foo implements IWithMethodImplementation1, IWithMethodImplementation2{
        @Override
        public String foo() {
            return IWithMethodImplementation1.super.foo();
        }
    }

    @Test
    public void WithMultiMethodImplementation() throws Exception {
        Foo foo = new Foo();
        assertThat(foo.foo()).isEqualTo("foo1");
    }
}

File: kotlin/src/test/kotlin/cn/staynoob/trap/kotlin/InterfaceSpec.kt
package cn.staynoob.trap.kotlin

import org.junit.Test
import org.assertj.core.api.Assertions.*

class InterfaceSpec {

    interface IWithMethodImplementation1 {
        fun foo() = "foo1"
    }
    interface IWithMethodImplementation2 {
        fun foo() = "foo2"
    }

    @Test
    fun withMethodImplementation() {
        class Foo : IWithMethodImplementation1 {}
        assertThat(Foo().foo()).isEqualTo("foo1")
    }

    @Test
    fun withMultiMethodImplementation() {
        class Foo :
                IWithMethodImplementation1,
                IWithMethodImplementation2 {
            override fun foo() = super<IWithMethodImplementation1>.foo()
        }
        assertThat(Foo().foo()).isEqualTo("foo1")
    }
}